### PR TITLE
net: Fix alignment when accessing network packet headers

### LINKED ISSUE #16587: build failures with gcc 9.x
**Describe the bug**
warnings with new compilers that turn into errors when building using sanitycheck

```
In file included from /home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:16:
/home/nashif/Work/zephyrproject/zephyr/include/net/net_pkt.h: In function ‘net_pkt_set_src_ipv6_addr’:
/home/nashif/Work/zephyrproject/zephyr/include/net/net_pkt.h:824:9: error: taking address of packed member of ‘struct net_ipv6_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  824 |         &NET_IPV6_HDR(pkt)->src);
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c: In function ‘arp_hdr_check’:
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:537:32: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  537 |      net_ipv4_is_addr_loopback(&arp_hdr->src_ipaddr)) {
      |                                ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c: In function ‘net_arp_input’:
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:590:9: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  590 |         &arp_hdr->src_ipaddr,
      |         ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:602:30: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  602 |       net_ipv4_is_addr_mcast(&arp_hdr->src_ipaddr)) {
      |                              ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:608:42: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  608 |   addr = if_get_addr(net_pkt_iface(pkt), &arp_hdr->dst_ipaddr);
      |                                          ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:635:8: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  635 |        &arp_hdr->src_ipaddr,
      |        ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:655:27: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  655 |   if (net_ipv4_is_my_addr(&arp_hdr->dst_ipaddr)) {
      |                           ^~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/subsys/net/l2/ethernet/arp.c:657:8: error: taking address of packed member of ‘struct net_arp_hdr’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  657 |        &arp_hdr->src_ipaddr,
      |        ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[3]: *** [zephyr/subsys/net/l2/ethernet/CMakeFiles/subsys__net__l2__ethernet.dir/build.make:63: zephyr/subsys/net/l2/ethernet/CMakeFiles/subsys__net__l2__ethernet.dir/arp.c.obj] Error 1
```
**To Reproduce**

Build for native_posix on Fedora 30 with new compiler

### PR DESCRIPTION
**Pull request by jukkar**
Avoid direct access to __packed struct elements. The structs
are typically cast to data that is received from the network.

Fixes #16587

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by jukkar** [comment]
This needs some more TLC. @rmstoi can you run this one against Maxwell?
There is a known issue as the ipv6_fragment test crashes in qemu_x86.

**Comment by rmstoi** [comment]
> This needs some more TLC. @rmstoi can you run this one against Maxwell as there are lot of small changes and it is difficult to manually test if anything fails.
> There is a known issue as the ipv6_fragment test crashes in qemu_x86, I will investigate it more.

Maxwell tests did not show regressions with this PR.

File: cmake/compiler/gcc/target_warnings.cmake [hunk 1]
```diff
@@ -14,8 +14,6 @@ macro(toolchain_cc_warning_base)
 
 if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "9.1.0")
   zephyr_compile_options(
-    # FIXME: Remove once #16587 is fixed
-    -Wno-address-of-packed-member
```
Comment (SebastianBoe) [inline cmake/compiler/gcc/target_warnings.cmake]: We shouldn't have an empty zephyr_compile_options() invocation. The whole if-statement should be removed.

**Review by SebastianBoe — DISMISSED** [review]

File: cmake/compiler/gcc/target_warnings.cmake (hunk 1 above)
Comment (jukkar) [reply cmake/compiler/gcc/target_warnings.cmake]: removed

File: include/net/net_if.h [hunk 2]
```diff
@@ -821,25 +821,56 @@ int net_if_config_ipv6_put(struct net_if *iface);
...
+/**
+ * @brief Check if this IPv6 address belongs to one of the interfaces.
+ *
+ * @param addr IPv6 address
+ * @param iface Pointer to interface is returned
+ *
+ * @return Pointer to interface address, NULL if not found.
+ */
+static inline
+struct net_if_addr *net_if_ipv6_addr_lookup_by_value(
```
Comment (pfalcon) [inline include/net/net_if.h]: This is weird signature. Why not return `struct net_if *` directly, out of which address can be gotten as needed, and get rid of "out" params?

**Comment by jukkar** [comment]
Fixed merge conflicts

**Comment by jukkar** [comment]
This PR feels "wrong" in many places in the diffs, and we should probably think this a bit more (-> adding DNM label). I just wonder if gcc is reporting proper things all the time.

**Comment by jukkar** [comment]
This was a failed experiment -> closing.

**Pull request closed without merging**